### PR TITLE
Refresh application routing and add tests

### PR DESCRIPTION
**Pull request by srtfisher**
- Add `without_middleware()` support.
- Bring REST API routing into parity with the rest of the routing logic.
- Allow `prefix()` and `name()` to work when prefixing the name/prefix of the route.
- Fix HEAD requests.
- Add tests for URL generation.
- Adjust URL generation to make it generate a proper route URL for a named route.

### CODE REVIEW

File: src/mantle/http/routing/class-router.php [hunk 1]
```diff
@@ -324,7 +356,20 @@ function( Request $request ) use ( $route ) {
...
 		// Ensure the response is valid since the middleware can modify it after it is run through Route.
-		return Route::ensure_response( $response );
+		return static::to_response( $request, $response );
+	}
+
+	/**
+	 * Prepare a response for sending.
+	 *
+	 * @param Request $request
+	 * @param mixed   $response
```
Comment (renatonascalves) [inline src/mantle/http/routing/class-router.php:366]: Missing a proper description.

File: src/mantle/http/routing/class-router.php [hunk 2]
```diff
@@ -492,33 +564,35 @@ public function substitute_implicit_bindings( Request $request ) {
...
+	 * @param string                $namespace Namespace for the REST API route.
+	 * @param callable|string       $callback  Callback that will be invoked to register
+	 *                                         routes OR a string route.
+	 * @param callable|array|string $args      Callback for the route if $callback is a
+	 *                                         string route OR arguments to pass to
+	 *                                         the register_rest_route() call. Not used if $callback
+	 *                                         is a closure.
 	 */
-	public function rest_api( string $namespace, $route, $args = [] ) {
+	public function rest_api( string $namespace, callable|string $callback, callable|array|string $args = [] ) {
```
Comment (renatonascalves) [inline src/mantle/http/routing/class-router.php:575]: Missing `@return`phpDoc and return type.

File: src/mantle/http/routing/class-url-generator.php [hunk 3]
```diff
@@ -211,23 +256,43 @@ public function get_root_url(): string {
...
 	}
 
+	/**
+	 * Get the base URL for the request.
+	 *
+	 * @param  string      $scheme
+	 * @param  string|null $root
+	 * @return string
+	 */
+	public function format_root( string $scheme, ?string $root = null ) {
```
Comment (renatonascalves) [inline src/mantle/http/routing/class-url-generator.php:275]:
```suggestion
	public function format_root( string $scheme, ?string $root = null ): string {
```

File: src/mantle/http/routing/class-url-generator.php [hunk 4]
```diff
@@ -250,7 +315,7 @@ public function is_valid_url( $path ): bool {
 	 * @param  string $scheme
 	 * @return void
 	 */
-	public function force_scheme( $scheme ) {
+	public function force_scheme( string $scheme ) {
```
Comment (renatonascalves) [inline src/mantle/http/routing/class-url-generator.php:318]:
```suggestion
	public function force_scheme( string $scheme ): void {
```

File: src/mantle/http/routing/class-url-generator.php [hunk 5]
```diff
@@ -250,7 +315,7 @@ public function is_valid_url( $path ): bool {
 	 * @param  string $scheme
 	 * @return void
```
Comment (renatonascalves) [inline src/mantle/http/routing/class-url-generator.php:316]: See my previous suggestion about `void`.

File: src/mantle/support/traits/trait-enumerates-values.php [hunk 6]
```diff
@@ -188,23 +188,21 @@ function ( $item ) use ( $key, $value ) {
 	 * @return void
 	 */
 	public function dd( ...$args ) {
```
Comment (renatonascalves) [inline src/mantle/support/traits/trait-enumerates-values.php:190]:
```suggestion
	public function dd( ...$args ): void {
```

File: src/mantle/support/traits/trait-enumerates-values.php [hunk 7]
```diff
@@ -188,23 +188,21 @@ function ( $item ) use ( $key, $value ) {
...
+		exit( 1 );
 	}
 
 	/**
 	 * Dump the items.
 	 *
-	 * @return $this
+	 * @return static
 	 */
 	public function dump() {
```
Comment (renatonascalves) [inline src/mantle/support/traits/trait-enumerates-values.php:201]:
```suggestion
	public function dump(): static {
```

File: src/mantle/testing/class-test-case.php [hunk 8]
```diff
@@ -39,6 +39,8 @@
 
 /**
  * Root Test Case for Mantle sites.
+ *
+ * Not designed for external use. Use {@see Mantle\Testkit\Test_Case} instead.
```
Comment (renatonascalves) [inline src/mantle/testing/class-test-case.php:43]: I can see a dev using this by mistake, especially if they have some IntelliSense in the editor where while importing the class, this note will not appear.

Would you consider adding this as part of custom namespace? Something like `Mantle\Testkit\Internal\Test_Case` or `Internal_Test_Case`...

**Review by renatonascalves — APPROVED** [review]
🗡️

File: src/mantle/testing/class-test-case.php (hunk 8 above)
Comment (srtfisher) [reply src/mantle/testing/class-test-case.php:43]: I can, I will rename it in a follow up

**Pull request merged**